### DISCUSSION TITLE
Is it possible to iterate through the same process multiple times using CrewAI

### DISCUSSION THREAD

**Discussion by okoliechykwuka** [crewAIInc/crewAI#1132] [Q&A]
I am using Crew to automate the writing of a proposal response for RFI documents.

I am writing my responses by following a ToC

```Table Of Contents:
-------------------------
1.0 Introduction

1.1 Background and objectives

1.2 Scope and deliverables

1.3 Project team and management structure
```
----------------------------------------------------------------
I have a lot of Agents handling this, The proposal Manager is a Super Agent coordinating the writing process.

I have written  the first part 1.0 Introduction, How do I instruct the Proposal Manager Agents that once we have the final output for 1.0, move to the next section 1.1

Can this be achieved using CrewAI?

I understand that the current process is Process.sequential, can It handle such case?

-------------------------------------------------------------------------------------------------

This is what my Task for the PM looks like.

```
    def navigate_table_of_content(self, agent):
        return Task(
            description=dedent(f"""
                The proposal's Table of Contents (TOC) is provided for your reference.
                
                ```{self.__toc()}```
                
                Remember, you are part of a team of agents working together on this task.
                
                YOUR TASK DETAILS:
                
                1. Inform the team about the current section we are focusing on, for example, '1.0 Introduction'.
                
                2. Provide a brief report on the current section or subsection number along with the title of that section.
                - For instance, 
                 "CURRENT SECTION  '1.0 Project Overview.'"
                 DON'T ADD ANYTHING ELSE TO THE REPORT.
                
                
                3. Proceed to the Next Section of the ToC Once the `Senior Proposal Writer` IS DONE WITH CURRENT SECTION. 
                
                {self.__tip_section()}
            """),
            agent=agent
        )
        ````

**Comment by C-1pher** [comment]
While I don't know if it possible using crewai itself , but one way to do this is to loop over the process using for loop  .

to do this make the prompt of the task (the description ) a variable and set the initial value to your prompt  that you used above.

in the loop execute the process and save the output in a result variable and add it to the prompt value so you can tell your agent at what stage of your ToC it's so it understand the context and what to do next .

**Comment by Biancamazzi** [comment]
don't forget we have https://chat.openai.com/g/g-qqTuUWsBY-crewai-assistant to help and also the discord community https://discord.com/invite/X4JWnZnxPb

**Comment by okoliechykwuka** [comment]
@sc00rpi0n I will try and apply the approach you suggested, I was thinking same appraoch

**Comment by iplayfast** [comment]
I had an example where the crew kickoff would happen as normal, and then a second kickoff would happen with the main agent returning "DONE" when the task was complete. A while loop on the results re kickedoff over and over.

**Comment by Kolaposki** [comment]
@iplayfast  Can you elaborate more on this or provide a little code snippet? Thanks!

**Comment by Kolaposki** [comment]
Check out the ``task.execute()`` method.
Thats what I use to get the output of a task and then move on to the next task or perform some operations. You can also use task.output which simply just gives you the result of the executed task.

**Comment by rodrigofarias-MECH** [comment]
> I had an example where the crew kickoff would happen as normal, and then a second kickoff would happen with the main agent returning "DONE" when the task was complete. A while loop on the results re kickedoff over and over.

A small snippet would be very useful indeed. Thanks!

**Comment by ATaylorAerospace** [comment]
@okoliechykwuka 
It sounds like you are looking for the kickoff_for_each() method. Check out the documentation to see if it is what you need. https://docs.crewai.com/core-concepts/Crews/?h=kickoff_for_each%28%29#kicking-off-a-crew

**Comment by EdFife** [comment]
Investigate the agent_name.execute_task(task_name, topic) functionality. I had some success with it.  
I could not make kickoff for each work.
I write out a bit more detail on my github page. https://github.com/EdFife/CrewAI-Looping/blob/main/README.md 

but here is my loop function:
for topic in inputs:
         # Initialize and execute the agent with the current topic
        if not topic:  # Skip invalid or empty topics
            print(f"Skipping invalid topic: {topic}")
            continue
        result = draft_writer.execute_task(chapter_writing_task, topic)  # draft_writer is my agent, chapter_writing_task is my task. Topic is the current list item to be written about. This line will actually launch the agent configurator in your terminal
